pkg/identityserver/end_device_registry.go

```diff
@@ -228,6 +228,37 @@ func (is *IdentityServer) updateEndDevice(ctx context.Context, req *ttnpb.Update
 	return dev, nil
 }
 
+func (is *IdentityServer) batchUpdateEndDevice(ctx context.Context, req *ttnpb.BatchUpdateEndDeviceRequest) (*ttnpb.EndDevices, error) {
+	if err := clusterauth.Authorized(ctx); err != nil {
+		return nil, err
+	}
+	req.FieldMask = cleanFieldMaskPaths([]string{"activated_at", "locations", "last_seen"}, req.FieldMask, nil, getPaths)
+	if len(req.FieldMask.GetPaths()) == 0 {
+		req.FieldMask = &pbtypes.FieldMask{Paths: updatePaths}
+	}
+	endDevices := make([]*ttnpb.EndDevice, len(req.EndDevices.EndDevices))
+	var fieldMask *pbtypes.FieldMask
+	err := is.store.Transact(ctx, func(ctx context.Context, st store.Store) (err error) {
+		for _, dev := range req.EndDevices.EndDevices {
+			*fieldMask = *req.FieldMask
+			if ttnpb.HasAnyField(req.FieldMask.GetPaths(), "activated_at") && dev.ActivatedAt == nil {
+				// The end device activation state may not be unset once set.
+				fieldMask = cleanFieldMaskPaths(ttnpb.EndDeviceFieldPathsNested, fieldMask, nil, []string{"activated_at"})
+			}
+			dev, err = st.UpdateEndDevice(ctx, dev, fieldMask.GetPaths())
+			if err != nil {
+				return err
+			}
+			endDevices = append(endDevices, dev)
+		}
+		return nil
+	})
```

Comment:
This isn't really a batch update. It will just be a large number of individual updates in a single transaction, where each update still requires a round-trip to the database. A real batch update would look more like:

```sql
WITH updates (application_id, device_id, last_seen_at) AS (
  VALUES
    ('foo', 'bar', '2022-03-08T...'),
    ('baz', 'qux', '2022-03-08T...'),
    ...
)
UPDATE end_devices AS end_device
SET last_seen_at = updates.last_seen_at
FROM updates
WHERE end_device.application_id = updates.application_id
AND end_device.device_id = updates.device_id
```

Reply:
Did it like this: https://github.com/TheThingsNetwork/lorawan-stack/blob/2ef3388754d6aa6009f30da8f59b4d570cd58dd1/pkg/identityserver/gormstore/end_device_store.go#L232-L263

It looks super hacky but I am afraid there is no better way. GORM (especially v1) doesn't really provide a nice way to do this, which is confusing to me because it seems like a common operation. Maybe experimenting with creating a temporary `Table` but I am not sure it's worth it. 

In Gorm V2 there is an option of inserting and then updating on conflict, but this is also something we don't want to do as we don't want to create entries for possibly deleted devices. Also in this way we can add extra restrictions (such as that the last seen timestamp in the db must be older than the one we want to replace it with).

At this point I think we should probably specialize this method for LastSeen updates only, but I left some room for whitelisting as @adriansmares suggested. There should be a way of adding other WHERE clauses as well.